pil-analyzer/src/type_inference.rs

```diff
@@ -624,22 +623,59 @@ impl<'a> TypeChecker<'a> {
             }
             Expression::BlockExpression(_, BlockExpression { statements, expr }) => {
                 let original_var_count = self.local_var_types.len();
+                let empty_tuple = Type::Tuple(TupleType { items: vec![] });
+                let expected_empty_tuple = ExpectedType {
+                    ty: empty_tuple.clone(),
+                    allow_array: false,
+                };
                 for statement in statements {
                     match statement {
                         StatementInsideBlock::LetStatement(LetStatementInsideBlock {
                             pattern,
                             value,
                         }) => {
                             let value_type = if let Some(value) = value {
-                                self.infer_type_of_expression(value)?
+                                let statement_type = self.infer_type_of_expression(value)?;
```

Comment:
Wait I just realize that this is the case where we have `let pattern = value;`. This means there should not be any change at all. We only need a change for `value;`-statements. And I think without a grammar change, they can only occur at the top level currently.